### DISCUSSION TITLE
syslog ingest pipeline sorely lacking ability to parse common syslog formats

### DISCUSSION THREAD

**Discussion by abesinger** [Security-Onion-Solutions/securityonion#5251] [Unsupported Versions] [answered]
Are the elasticsearch ingest pipelines written/maintained by SecurityOnion, or are they just imported from Elastic? If it makes any difference, I'm specifically asking about the syslog pipeline. I'd like to know where I should be posting bug reports on them.

I just started forwarding syslog from a RHEL host (rsyslog forwarding to filebeat on a sensor node).  Rsyslog is using a very common message format, but Elastic is doing a lousy job of parsing fields, e.g. host.name is getting set to the name of the sensor node, not the hostname from the syslog message. 

Looking at the syslog ingest pipeline, it's extremely limited in what it knows how to parse -- it only has patterns for two formats, one that matches messages without a hostname (!), and the other matches the (not so) Common Event Format logs. So, basically, unless you're importing messages in one of those two formats you're not going to have the  most useful data.

Makes me wonder how much people are using SO to look at syslog messages.

I can (and will) develop some addition grok patterns to handle some more common formats. But they should get merged into the distribution so that I don't have to worry about updates overwriting things.

So, should I be posting about it here, or to Elastic?

**Comment by strf0x1** [comment]
I agree, there could be a lot improved about how linux logs are parsed out of the box. SO seems to rely more on Wazuh from a security perspective. But there is a nice alternative now as of 2.3.6x. Elastic's best parsers (in the form of ingest node pipelines) are available via their Filebeat modules. SO now supports them via Salt https://docs.securityonion.net/en/2.3/filebeat.html.  You can take a look  at the modules https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-modules.html. Maybe what you're looking for is the System module? So you would basically setup the Filebeat container to enable the module you want. This would create the ingest node pipeline on ElasticSearch. Then you can install Filebeat on the servers you wish to log on, point them at Logstash and you should be good to go. At least maybe this will give you a start.

**Comment by abesinger** [comment]
Thanks for the response! (although I'd still like to know who owns the ingest parsers)

Your suggestion is a nice idea, but...

I did look at the system module, the problem is that, unless I'm missing something, the system module only works on local (log) files, not logs received over TCP. In my case I'm using  TCP  input,  the source hosts are sending using rsyslog. At least, the documentation for the system module only mentions file paths.

I'm not looking to install filebeat on all of the hosts that are generating logs. I've hundreds of existing hosts; it's a lot easer to ask sysadmins to make an addition to rsyslog.conf than to install and support and additional piece of software.

I suppose I could work around the system moduloe limits by doing a store-and-forward -- using rsyslog as the receiver on the filebeat host, store the logs in local files, and the use filebeat to ship them to SO. That has its pluses and minuses.

**Comment by strf0x1** [comment]
OK, I think I see what you're doing... have done this as well with a central rsyslog collector. I think you can still make it work with one Filebeat instance running on the central rsyslog server. Filebeat system module allows you to define the path to log files, and supports recursive directories. So your central rsyslog server could store them something akin to /var/log/remotelogs/system/host1/. Haven't tested it extensively and it's not perfect, but think it could definitely work. One downside would be that your hostname field will actually show the central syslog aggregator, but there should be an additional field that shows the actual source host. This would prob break default dashboards, but not too hard to fix. Would be curious to hear if you find a solution.

**Reply by abesinger** [reply]
Well, you just convinced me _**not**_ to bother trying the filebeat syslog module.  If it can't get the hostname out of the message, it's not worth the bother. &lt;rant> It's sad, I literally [wrote the book](https://www.usenix.org/short-topics/building-logging-infrastructure)  on this 17 years ago,  haven't been doing some much with logs recently, but I see preserving hostnames continues to get overlooked. What good is a log message when you don't know what host it came from? &lt;/rant>

So I'm going to keep going in the direction I was initially going, which is to modify the Elastic syslog ingest pipeline to hand the message format coming from rsyslog. It's mostly a matter of adding a grok pattern and setting a few fields.  I've already pretty much got it done, just need to test. I'll post the results once it's working.

**Reply by wdpless** [reply]
I'd love to see it when your done.

**Reply by abesinger** [reply]
And here it is.  I took the existing `syslog` pipeline,  added a grok pattern to match rsyslog's `Rsyslog_FWDFORMAT` template, and added some "set" commands to copy the syslog fields to ECS compliant fields.  And for giggles, I added a painless script that breaks out the syslog priority code into facility and level.

And I cleaned up the formatting of the rest of the pipeline so that it's more readable.

Unfortunately the painless code is a bit difficult to read, as it has to be on a single line; I had a nice bit of well formatted, commented  code that worked using Elastic's Dev Tool, but multi-line quoting (`"""`) only works through the dev interface,  doesn't work when SO uploads the code. 

```
{
  "description" : "syslog",
  "processors" : [
    {
      "dissect": {
        "field": "message",
        "pattern" : "%{message}",
        "on_failure": [ { "drop" : { } } ]
      },
      "remove": {
        "field": [ "type", "agent" ],
        "ignore_failure": true
      }
    }, {
      "grok": {
        "field": "message",
          "patterns": [
            "^<%{INT:syslog.priority:int}>%{TIMESTAMP_ISO8601:syslog.timestamp} +%{IPORHOST:syslog.host} +%{PROG:syslog.program}(?:\\[%{POSINT:syslog.pid:int}\\])?: %{GREEDYDATA:real_message}$",

            "^<%{INT:syslog.priority}>%{DATA:syslog.timestamp} %{WORD:source.application}(\\[%{DATA:pid}\\])?: %{GREEDYDATA:real_message}$",

            "^%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.host} %{SYSLOGPROG:syslog.program}: CEF:0\\|%{DATA:vendor}\\|%{DATA:product}\\|%{GREEDYDATA:message2}$"
          ],
          "ignore_failure": true
      }
    }, {
      "script": {
        "description": "Map syslog priority into facility and level",
        "lang": "painless",
        "params" : {
          "level": [
            "emerg",
            "alert",
            "crit",
            "err",
            "warn",
            "notice",
            "info",
            "debug"
           ],
           "facility" : [
             "kern",
             "user",
             "mail",
             "daemon",
             "auth",
             "syslog",
             "lpr",
             "news",
             "uucp",
             "cron",
             "authpriv",
             "ftp",
             "ntp",
             "security",
             "console",
             "solaris-cron",
             "local0",
             "local1",
             "local2",
             "local3",
             "local4",
             "local5",
             "local6",
             "local7"
            ]
        },
        "source": "if (ctx['syslog'] != null && ctx['syslog']['priority'] != null) { int p = ctx['syslog']['priority']; int f = p / 8; int l = p - (f * 8); ctx['syslog']['facility'] = [ : ]; ctx['syslog']['level'] = [ : ]; ctx['syslog']['level'].put('code', l); ctx['syslog']['level'].put('name', params.level[l]); ctx['syslog']['facility'].put('code', f); ctx['syslog']['facility'].put('name', params.facility[f]); }"
      }
    }, {
      "set": {
        "if": "ctx.syslog?.host != null",
        "field": "host.name",
        "value": "{{ syslog.host }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.syslog?.program != null",
        "field": "process.name",
        "value": "{{ syslog.program }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.syslog?.pid != null",
        "field": "process.id",
        "value": "{{ syslog.pid }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.source?.application == 'filterlog'",
        "field": "dataset",
        "value": "firewall",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.vendor != null",
        "field": "module",
        "value": "{{ vendor }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.product != null",
        "field": "dataset",
        "value": "{{ product }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "field": "ingest.timestamp",
        "value": "{{ @timestamp }}"
      }
    }, {
      "date": {
        "if": "ctx.syslog?.timestamp != null",
        "field": "syslog.timestamp",
        "target_field": "@timestamp",
        "formats": ["MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601", "UNIX"],
        "ignore_failure": true
      }
    }, {
      "remove": {
        "field": ["pid", "program"],
        "ignore_missing": true,
        "ignore_failure": true
      }
    }, {
      "pipeline": {
        "if": "ctx.vendor != null && ctx.product != null",
        "name": "{{ vendor }}.{{ product }}",
        "ignore_failure": true
      }
    }, {
      "pipeline": {
        "if": "ctx.dataset == 'firewall'",
        "name": "filterlog",
        "ignore_failure": true
      }
    }, {
      "pipeline": { "name": "common" }
    }
  ]
}

```

**Reply by strf0x1** [reply]
That's really cool, thanks for sharing.

**Reply by dylangerdaly** [reply]
How can I install this?

**Reply by abesinger** [reply]
Copy it into `/opt/so/saltstack/local/salt/elasticsearch/files/ingest/syslog` on your manager node, and let `salt` do its thing. It will appear as `/opt/so/conf/elasticsearch/ingest/syslog` on your search nodes.  That's it!

Be aware that this local file will overwrite the default syslog pipeline file; if SecurityOnion distributes a newer version of the pipeline in a future release, your version will continue to overwrite the new version and you'll need to manually check and resolve any differences.

**Comment by abesinger** [comment, marked as answer]
Reposting my response to the thread above so it appears as an answer:

And here it is.  I took the existing `syslog` pipeline,  added a grok pattern to match rsyslog's `Rsyslog_FWDFORMAT` template, and added some "set" commands to copy the syslog fields to ECS compliant fields.  And for giggles, I added a painless script that breaks out the syslog priority code into facility and level.

And I cleaned up the formatting of the rest of the pipeline so that it's more readable.

Unfortunately the painless code is a bit difficult to read, as it has to be on a single line; I had a nice bit of well formatted, commented  code that worked using Elastic's Dev Tool, but multi-line quoting (`"""`) only works through the dev interface,  doesn't work when SO uploads the code. 

```
{
  "description" : "syslog",
  "processors" : [
    {
      "dissect": {
        "field": "message",
        "pattern" : "%{message}",
        "on_failure": [ { "drop" : { } } ]
      },
      "remove": {
        "field": [ "type", "agent" ],
        "ignore_failure": true
      }
    }, {
      "grok": {
        "field": "message",
          "patterns": [
            "^<%{INT:syslog.priority:int}>%{TIMESTAMP_ISO8601:syslog.timestamp} +%{IPORHOST:syslog.host} +%{PROG:syslog.program}(?:\\[%{POSINT:syslog.pid:int}\\])?: %{GREEDYDATA:real_message}$",

            "^<%{INT:syslog.priority}>%{DATA:syslog.timestamp} %{WORD:source.application}(\\[%{DATA:pid}\\])?: %{GREEDYDATA:real_message}$",

            "^%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.host} %{SYSLOGPROG:syslog.program}: CEF:0\\|%{DATA:vendor}\\|%{DATA:product}\\|%{GREEDYDATA:message2}$"
          ],
          "ignore_failure": true
      }
    }, {
      "script": {
        "description": "Map syslog priority into facility and level",
        "lang": "painless",
        "params" : {
          "level": [
            "emerg",
            "alert",
            "crit",
            "err",
            "warn",
            "notice",
            "info",
            "debug"
           ],
           "facility" : [
             "kern",
             "user",
             "mail",
             "daemon",
             "auth",
             "syslog",
             "lpr",
             "news",
             "uucp",
             "cron",
             "authpriv",
             "ftp",
             "ntp",
             "security",
             "console",
             "solaris-cron",
             "local0",
             "local1",
             "local2",
             "local3",
             "local4",
             "local5",
             "local6",
             "local7"
            ]
        },
        "source": "if (ctx['syslog'] != null && ctx['syslog']['priority'] != null) { int p = ctx['syslog']['priority']; int f = p / 8; int l = p - (f * 8); ctx['syslog']['facility'] = [ : ]; ctx['syslog']['level'] = [ : ]; ctx['syslog']['level'].put('code', l); ctx['syslog']['level'].put('name', params.level[l]); ctx['syslog']['facility'].put('code', f); ctx['syslog']['facility'].put('name', params.facility[f]); }"
      }
    }, {
      "set": {
        "if": "ctx.syslog?.host != null",
        "field": "host.name",
        "value": "{{ syslog.host }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.syslog?.program != null",
        "field": "process.name",
        "value": "{{ syslog.program }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.syslog?.pid != null",
        "field": "process.id",
        "value": "{{ syslog.pid }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.source?.application == 'filterlog'",
        "field": "dataset",
        "value": "firewall",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.vendor != null",
        "field": "module",
        "value": "{{ vendor }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.product != null",
        "field": "dataset",
        "value": "{{ product }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "field": "ingest.timestamp",
        "value": "{{ @timestamp }}"
      }
    }, {
      "date": {
        "if": "ctx.syslog?.timestamp != null",
        "field": "syslog.timestamp",
        "target_field": "@timestamp",
        "formats": ["MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601", "UNIX"],
        "ignore_failure": true
      }
    }, {
      "remove": {
        "field": ["pid", "program"],
        "ignore_missing": true,
        "ignore_failure": true
      }
    }, {
      "pipeline": {
        "if": "ctx.vendor != null && ctx.product != null",
        "name": "{{ vendor }}.{{ product }}",
        "ignore_failure": true
      }
    }, {
      "pipeline": {
        "if": "ctx.dataset == 'firewall'",
        "name": "filterlog",
        "ignore_failure": true
      }
    }, {
      "pipeline": { "name": "common" }
    }
  ]
}

```

**Reply by revast** [reply]
Just FIY, the default Syslog Parser config **[/opt/so/conf/elasticsearch/ingest/syslog](https://raw.githubusercontent.com/Security-Onion-Solutions/securityonion/37251301284ad6560dc324018ab85a27336853eb/salt/elasticsearch/files/ingest/syslog)** , which this Custom Config is based on, seems to have changed (right side is default):

![image](https://user-images.githubusercontent.com/8189198/159904005-ec6f8f6e-ebe4-4336-9215-f1e4e9b19a3e.png)
